render_task: Assign passes in a single pass.

emilio opened the pull request:
Right now creating render passes is a three-step process:

 * You first go through the render task tree and figure out the max depth.
 * You allocate as many passes as needed.
 * You traverse the whole tree again assigning to those passes.

This patch makes it so that assign_to_passes is responsible to grow the passes
vector with offscreen passes lazily as needed, and its caller is responsible to
reverse the array so that the passes end up in order.

Thus so that we can build the pass list in a single pass (no pun intended) as we
traverse down the tree.

This also removes the SpecialRenderPasses because now that cached render tasks
know how to create passes as needed I'm pretty sure it's not needed. There's no
dependency between the two passes I could think of, looking at the code, and all
the tasks use color render targets, so I think they're unnecessary.

And in any case they should be easy to add back, should pathfinder really needed
them.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/3445)
<!-- Reviewable:end -->

emilio commented:
r? @gw3583

emilio commented:
Try is closed due to infra issues. I pushed regardless just in case it works so I don't need to do it in the morning:

https://treeherder.mozilla.org/#/jobs?repo=try&revision=af56103f937386bda6d3189f30d4652399548a8f

Otherwise I'll push in the morning and ensure this is green before landing.

emilio commented:
(Assuming it's r+'d of course :P)

emilio commented:
Try looks good.

kvark commented:
> This also removes the SpecialRenderPasses because now that cached render tasks
know how to create passes as needed I'm pretty sure it's not needed.

I'm slightly confused by this (haven't gone through the code yet). The tasks should be creating passes in the first place, since passes are higher level concept.

>  There's no dependency between the two passes I could think of, looking at the code, and all
the tasks use color render targets, so I think they're unnecessary.

Please clarify this part. Things are separated into passes, conceptually, so that *with a pass* there are no dependencies. But dependencies between passes are naturally expected. Also, we do have alpha render targets (as opposed to color), and even pathfinder's special tasks have alpha, as far as I see.

emilio commented:
> I'm slightly confused by this (haven't gone through the code yet). The tasks should be creating passes in the first place, since passes are higher level concept.

So these passes were passes specific to pathfinder, where all the glyphs were rendered before-hand. This was, I think, working around the bug I fixed in #3444 (which is much more likely to occur with glyphs). Now the more generic code that I added in that PR should handle it, by creating passes as necessary.

> Please clarify this part. Things are separated into passes, conceptually, so that with a pass there are no dependencies. But dependencies between passes are naturally expected. Also, we do have alpha render targets (as opposed to color), and even pathfinder's special tasks have alpha, as far as I see.

Yes, I understand that. My point is that I don't think there's any dependency between `special_render_passes.alpha_glyph_pass` and `special_render_passes.color_glyph_pass`. And as such there shouldn't be any reason for their existence, they should just be merged with the rest of the `cacheable_render_tasks` passes.

kvark commented:
Thanks for clarification!
@bors-servo r+

bors-servo commented:
:pushpin: Commit 407e244 has been approved by `kvark`

<!-- @bors-servo r=kvark 407e244253be1b7c9e37fa87398fc375c55358f1 -->

bors-servo commented:
:hourglass: Testing commit 407e244253be1b7c9e37fa87398fc375c55358f1 with merge 5da3e6bf916e84fc3ce72c6cf65c63df0cdd8eb5...

emilio commented:
> shouldn't we leave unimplemented! here, given that the original passes no longer exist?

No, because this should use the more generic mechanism described above.

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/builds/21218220), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/AEmiPV3KSxKwP5nGSjFRkw)
Approved by: kvark
Pushing 5da3e6bf916e84fc3ce72c6cf65c63df0cdd8eb5 to master...
